### ISSUE TITLE
Incredible new results!

### ISSUE THREAD

**Issue by melgor** [davidsandberg/facenet#152]
Congratulation about new results. I've a couple of question about the results:

1. About results on CASIA-WebFace: did you use clean_list or all images from CASIA?
2. About results on MS-Celeb:
 -  as I understand, you filter the data using your own method. Could you release the clean_list or maybe you do this online, so there is no such thing?
 - what about learning parameters, there are the same like in the [Wiki](https://github.com/davidsandberg/facenet/wiki/Classifier-training-of-inception-resnet-v1)?
 -  how many images do you have in data and how long does it take to learn the network?
 - what was your final accuracy at MsCeleb data?

Again, congratulation for a really good results!

**Comment by davidsandberg** [comment]
Thanks @melgor!
1. For the CASIA results no additional cleaning is used, but I use the [casia-maxpy-clean](https://groups.google.com/forum/#!topic/cmu-openface/Xue_D4_mxDQ) version.
2. For the MS-Celeb-1M results I cleaned the dataset by selecting a subset of the training images based on the distance for each image to the class center. I plan to write a wiki page for how I did it when I get some time. I didn't use a clean list as such, but instead had a file containing the distance between that image embedding to its class center. And when running training I can specify decide to use only the 75% of the images that are closest to its class center. 
Learning parameters are different as well. In #48 you can see the learning curves with some hyper parameters in the legend (wd=weight decay, cl=center loss, Kp=keep percentile, 75 => keep only the 75% of the images closest to its class centers). For all runs I removed classes with less than 60 images.
The number of images and classes depends on the settings, but for example for Kp=75 gave me 4 213 410 images over 51 261 classes. Training took ~43 hours (250 000 steps with batch size 90), but the learning rate schedule can probably be optimized quite a bit).
Final accuracy has varied between 0.993 and 0.995.

**Comment by scotthong** [comment]
Hi @davidsandberg:
Thanks for sharing this project and congratulation about new results!
I am trying to reproduce the results using the MS-Celeb-1M dataset. I couldn't find the the following two hyper parameters documented your post, and can you share the parameter you've used to achieve the results. 

--keep_probability
--center_loss_alfa

Thanks!

**Comment by ugtony** [comment]
Hi, @davidsandberg,
I used your msceleb1m model and calculating_filtering_metrics.py to get the distance_to_center values on msceleb1m dataset. 

I plotted the histogram of distance_to_center in order to get a better idea how keep_percentile can be set.

And then I found the histogram seems to be formed by two Gaussian models. The left one is centered around 0.6 and the right one is center around 1.0. 
Do you have any idea how to explain that?

Are the instances belong to the right Gaussian the noisy data? Or are they profile faces?
If the answer is the former one, then setting percentile 75 might be risky.
If the answer is the latter one, then center_loss didn't make the model do well on profile faces.
![default](https://cloud.githubusercontent.com/assets/6657593/23889853/10d025a2-08ca-11e7-87e6-a6e15e72620e.png)

**Comment by se7oluti0n** [comment]
@ugtony @davidsandberg in file `calculating_filtering_metrics.py `, you calculate `embeddings` features for each image. As I understand you do filtering images in training phase for MS-Celeb-1M. So which model you use in `calculating_filtering_metrics.py ` to calculate `embeddings` features? May be model trained with Casia dataset

**Comment by ljstrnadiii** [comment]
Hi @davidsandberg ,

I have this idea to use the embeddings in another model. I am building a VAE model to generate faces and I would like to use the embeddings of FaceNet to add a loss to the reconstruction of the decoder. I am thinking something like:

total_loss = tf.reduce_mean( (1-alpha)*reconstructionL2 + alpha*embeddingsL2)

Do you have any thoughts or concerns? I am trying to dig into your code to get an understanding of how I could generate the embeddings of the actual and generated images at the loss, but it is not that intuitive at the moment. 

Any suggestions on how to incorporate the embedding into a loss function?

cheers

**Comment by davidsandberg** [comment]
Hi @ljstrnadiii, 
That sounds like a very cool project. I'm looking forward to hear more about it.
You could for example start with train_softmax.py and use your total_loss above instead of the softmax_cross_entropy. But then you of course need to add the VAE parts as well.

**Comment by ljstrnadiii** [comment]
Hi @davidsandberg,

Right! I am excited to see what happends. Anyways, I was thinking of simply using a pretrained model to calculate the embeddings of the generated image and the actual image and using `tf.nn.l2_loss(.,.)` between the two. I am pretty new to tf. I have a few question and maybe you could help. It is a bit more conceptual:

It seems that I can just load the model in the default graph like so:

        self.fn_saver = tf.train.import_meta_graph(ps.path.join(self.fn_model_exp,self.fn_meta_file))
        
        # load the tensors of facenet:
        embed_images_placeholder = tf.get_default_graph().get_tensor_by_name("input:0")
        # stop gradient to prevent training the FaceNet model.
        embeddings = tf.stop_gradient(tf.get_default().get_tensor_by_name("embeddings:0"))
        phase_train_placeholder = tf.get_default_graph().get_tensor_by_name("phase_train:0")

Then I can call 
        
        sess=tf.Session()
        self.fn_saver.restore(sess, os.path.join(self.model_exp,self.ckpt_file))
        gen_embed = sess.run(embeddings, feed_dict={embed_images_placeholder:gen_images, 
                                        phase_train_placeholder: False})

to get the embeddings of `gen_images`. The part that I am having a hard time grasping is that in my VAE training scheme in each iteration I call 

        sess.run((merged, self.optimizer), feed_dict=images)

but I need to intercept the optimizer graph flow to get the generated images to calculate the embeddings. The vae flow is optimizer <- cost <- latent loss(samples), gen loss(generate) <- generate(samples) <- samples(images). So, what I think I have to do is in the training function for the VAE I need to make multiple runs and break apart the process like so:

        gen_images = sess.run(generated_images, feed_dict=images)
        gen_embed = sess.run(embeddings, feed_dict={embed_images_placeholder:gen_images,
                                    phase_train_placeholder: False}
        real_embed = sess.run(embeddings, feed_dict={embed_images_placeholder:real_images, 
                                    phase_train_placeholder: False}

Then, I can treat these last two results as values for placeholders to compute the loss of run the optimizer:
   
        summary = sess.run((merged, self.optimizer), feed_dict={embed_real: real_embed, embed_gen: gen_embed, images:images})

in order to present the embeddings for the L2 embeddings loss. Then the plan is to simply sum the three losses and try to train. I realize this is a lot and I may be barking up the wrong tree, but am I getting the jist of what I should do? Does the idea make sense roughly or am I breaking some basic tf rules? I could go to stack exchange, but I think the explanation would be about three times longer!

**Comment by ljstrnadiii** [comment]
It looks like I would have to use something like tensorflow's sess.partial_run(). The only other way I could see it working is if I could somehow add the embeddings as an op or node in the graph without having to call sess.run(embeddings, feed_dict=feed_dict), but from what I can tell that is not an option.

**Comment by ljstrnadiii** [comment]
@davidsandberg ,

Thanks for the suggestion. I am pretty close to implementation. What is the file format of the pretrained model you restore in the softmax_train.py file? I am pointing the restore function to  the ckpt file in the pretrained model dir, but no luck. It says maybe the file is in a different format

**Comment by ljstrnadiii** [comment]
I believe it has to do with just restoring the variables in the inception_resnet_v1 model. Any ideas?
